accepted/cranelift-roadmap-2022.md

```diff
@@ -0,0 +1,895 @@
+recursion-based flow control into loops and other more efficient
+structures.
+
+### Exception Handling
+
+Another language feature that requires explicit compiler support is
+*exception handling*: the ability to unwind stack frames and perform
+cleanup along the way, stopping at relevant handlers, all without
+overhead in the common (non-exception-throwing) case.
+
+Zero-overhead exception support requires several things from the
+compiler. First, it must emit metadata that correlates locations in
+the program (PC values) to relevant handlers. It must also emit
+precise metadata that enables unwinding of stack frames from any point
+that might experience a thrown exception. Finally, exceptional control
+flow must be taken into account by optimizations.
+
+We will likely refer to existing designs here, for example [LLVM's
+exception
+handling](https://llvm.org/docs/ExceptionHandling.html). That said,
+our initial design should be driven by what is necessary to support
+the [Wasm exception handling
+proposal](https://github.com/WebAssembly/exception-handling), keeping
+in mind more general use-cases to ensure we make the design
+future-proof.
+
+We have [started to
+discuss](https://github.com/bytecodealliance/wasmtime/issues/3427)
+this topic recently; initially we had been converging on possibly a
+simple purpose-built unwinder, but after more discussion in [a
+recently Wasmtime biweekly
+meeting](https://github.com/bytecodealliance/wasmtime/blob/main/meetings/wasmtime/2021/wasmtime-10-28.md),
+it seems likely we will stick to platform-based support using DWARF
+(and eventually Windows Structured Exception Handling) for now.
+
+### Garbage Collection (GC)
+
+Currently, Cranelift provides the necessary support for precise GC
+rooting via "reference types": the `r32` and `r64` value types can be
+passed into Cranelift functions as arguments, propagate through the
+dataflow as opaque values, enter and leave specific kinds of storage
+(such as tables), and be returned. The register allocator and the
+lowering logic are careful to preserve referenced-typed-ness
+throughout the pipeline so that a GC can accurately identify which
+references are live.
+
+However, Cranelift does not yet provide adequate support for actually
+using or manipulating the references from within CLIF: there are
+[pending
+issues](https://github.com/bytecodealliance/wasmtime/issues/3217)
+where attempted casts between references and integers leads to
+regalloc panics. This should be supported: it should be possible to
+emit CLIF code that takes an `r32` or `r64`, casts it to an actual
+pointer (possibly using knowledge about the reference format used by
+the embedding runtime, e.g. with respect to tag bits or NaN boxing
+that need to be stripped) and use it. We should update the register
+allocator (likely just regalloc2) and provide canonical `ref_to_int`
+and `int_to_ref` operators that provide a supported and tested way to
+do this.
```

Comment:
I think we actually *don't* want to encourage the use of things like `ref_to_int` and `int_to_ref`. This will still either suffer from the "whoops, no stack map entry for this value because its an `int` and not a `ref`" problem, or we will have to start adding stack map entries for `int`s derived from a `ref`, but then where do we stop? What about `int`s derived from an `int` derived from a `ref`? It gets messy really quick. Instead, I think we want to allow using `ref`s as address operands for loads and stores, and maybe a few other places where `int` operands are currently allowed, but `ref`s aren't (maybe even make `ref` usable *everywhere* `int` is, basically making them exactly the same *except* that `ref`s get stack map entries).

But all of this is a bit too much in the weeds for a roadmap RFC. I think it suffices to say that we should make `r{32,64}` usable for non-opaque GC objects where we can load from/store to their fields (including fields that are themselves `r{32,64}`s) inline instead of only with out-of-line VM functions like what would be required now.

Reply:
Yeah, that's a good point. I guess I had been imagining the lowering of a field access to be a "cast to int, mask off the tag bits, and load, while carefully avoiding safepoints" sort of operation (because the ref is dead/unrooted after the cast). But that's a big footgun. First-class operators that use refs as pointers are almost certainly better in that regard. 

One consideration I had in mind is that the conversion from ref to actual pointer actually needs to be a little bit flexible to work in various runtimes: I'm aware of at least low-bit tagging (classical "pointers are aligned so use 2/3 tag bits"), NaN boxing or upper-bit tagging generally (top 16 bits), and the JVM's "compressed oop" design where a 32-bit pointer is used in 64-bit space with `base + oop << 3` (allowing a 2^35 / 32 GiB heap). We could in theory have a load_ref/store_ref that takes a mask, a shift, and an addend, but maybe there's something better.

Another possibility that occurs to me: allow casts to ints, and allow the user to generate CLIF as needed, but *also* take the ref itself as an arg to the load/store. This requires keeping the ref alive up to the point of use, at least at the CLIF level. (If the load/store isn't a safepoint for the particular runtime then the lowering can drop it.) So we'd have `int := ref_to_int r; actual_ptr := arbitrary_conversion(int); value := load_ref actual_ptr, r`.

Anyway, yeah, I'll update the RFC to be more generic on this point but this will be an interesting design discussion to have later!

Reply:
I’d like to add another point on the design space of deriving pointers from GC objects, although agreed that it warrants a deeper discussion elsewhere.

An issue SpiderMonkey will have with the Wasm-GC proposal when we add optimizing support is that some of our GC objects are represented as a GC ref to a cell which contains a regular heap pointer to the actual data. The outline data’s lifetime is tied to the GC object and is freed with a finalizer. When we access a field/element for an object then, we need to acquire the data pointer and then get the field/element from the data pointer.

There’s a potential hazard here however if the GC ref’s lifetime ends before the use of the data pointer and there is a safepoint/collection after the GC ref’s lifetime ends and before the use of the data pointer. This can’t be solved with just allowing r32/r64 to be manipulated as integers as the derived data pointer is not a ref that can be traced and shouldn’t get a stackmap entry. SM solves this by adding a synthetic use of the original GC ref to any loads using the data pointer. This extends the lifetime of the GC ref to match all uses of the data pointer. There may be other solutions here as well.

Additionally, the outline data representation for GC objects is a pain for codegen efficiency and ideally would be eliminated. But that will require some work in the SM GC that may not be feasible right away.

Reply:
Updated this description to note that there is a design space to explore, and some constraints (avoiding rooting issues, also being flexible across different reference representations, such as indirection or tagging approaches).